### PR TITLE
request_id: allow disabling trace reason packing

### PR DESCRIPTION
**Pull request by mattklein123**
1) Promote the default UUID request_id implementation to
   an actual extension that is required in the build and wire
   up all documentation.
2) Add a configuration option to the extension that allows trace
   reason packing to be disabled (the default continues to be for it
   to be enabled to match existing behavior).
3) Update all documentation for the new behavior.
4) Substantial cleanup of these code paths for clarity and robustness.

Fixes https://github.com/envoyproxy/envoy/issues/11532

Risk Level: Low
Testing: New/modified tests
Docs Changes: Added
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @euroelessar if you could take a look that would be great. I think this will cause some compilation breakages for your internal plugin but I think it's a worthwhile change overall.

File: test/extensions/filters/network/http_connection_manager/config_test.cc [hunk 1]
```diff
@@ -1708,8 +1707,30 @@ TEST_F(HttpConnectionManagerConfigTest, DefaultRequestIDExtension) {
                                      date_provider_, route_config_provider_manager_,
                                      scoped_routes_config_provider_manager_, http_tracer_manager_,
                                      filter_config_provider_manager_);
-  auto request_id_extension =
-      dynamic_cast<Http::UUIDRequestIDExtension*>(config.requestIDExtension().get());
+  auto request_id_extension = dynamic_cast<Extensions::RequestId::UUIDRequestIDExtension*>(
+      config.requestIDExtension().get());
+  ASSERT_NE(nullptr, request_id_extension);
```
Comment (dmitri-d) [inline test/extensions/filters/network/http_connection_manager/config_test.cc:1712]: nit: Here, and in `DefaultRequestIDExtensionWithParams` test: would be nice to be able to verify config particulars (even though it's just one parameter atm), not just that the extension has been instantiated.

**Comment by mattklein123** [comment]
@envoyproxy/windows-dev can one of you help me take a look at `//test/common/http:conn_manager_utility_test` in this PR? I tried to fix it with the last commit which seemed to fix the Windows clang build but not the msvc build, and the msvc error is not clear on what the problem is. Thank you.

**Comment by davinci26** [comment]
I can take a look at it tomorrow morning

**Comment by mattklein123** [comment]
> I can take a look at it tomorrow morning

Thanks but nevermind, I think I figured it out. It had to do with some compilers (including gcc) not liking a Mock to interface shared_ptr cast in an expectation.

File: test/extensions/filters/network/http_connection_manager/config_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1712]: Fixed

**Comment by mattklein123** [comment]
I'm struggling to see how I caused this clang-tidy issue but I can take a look. Otherwise I think this is ready to review.

File: api/envoy/extensions/request_id/uuid/v3/uuid.proto [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+// Configuration for the default UUID request ID extension which has the following behavior:
+//
+// 1. Request ID is propagated using the *x-request-id* header.
+//
+// 2. Request ID is a universally unique identifier (UUID).
+//
+// 3. Tracing decision (sampled, forced, etc) is set in 14th byte of the UUID. By default this will
+//    overwrite existing UUIDs received in the *x-request-id* header if the trace sampling decision
+//    is changed.
```
Comment (htuch) [inline api/envoy/extensions/request_id/uuid/v3/uuid.proto:25]: Is there any documentation on the format for the 14th byte to link to here? Is this just 1 or 0? Presumably the overwrite only happens if we are deciding to sample on a request ID that doesn't already indicate, vs. the other way around? Sorry if these aren't great questions, but as an unfamiliar reader, this is likely a question to arise.

Comment (mattklein123) [reply api/envoy/extensions/request_id/uuid/v3/uuid.proto:25]: I don't think there is but I can beef up the docs for this.

File: api/envoy/extensions/request_id/uuid/v3/uuid.proto [hunk 3]
```diff
@@ -16,13 +16,23 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+//    <config_http_conn_man_headers_x-request-id>` header.
 //
-// 2. Request ID is a universally unique identifier (UUID).
+// 2. Request ID is a universally unique identifier `(UUID4)
+//    <https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)>`_.
 //
 // 3. Tracing decision (sampled, forced, etc) is set in 14th byte of the UUID. By default this will
 //    overwrite existing UUIDs received in the *x-request-id* header if the trace sampling decision
-//    is changed.
+//    is changed. The 14th byte of the UUID4 has been chosen because it is fixed to '4' by the
```
Comment (mattklein123) [inline api/envoy/extensions/request_id/uuid/v3/uuid.proto:27]: Also this should really be s/byte/nibble I think. I can make this change in the next round of comments.

File: include/envoy/http/request_id_extension.h [hunk 4]
```diff
@@ -5,19 +5,37 @@
...
+
+  /**
+   * Perform a mod operation across the request id within a request and store the result in the
+   * provided output variable. This is used to perform sampling and validate the request ID.
+   * @param request_headers supplies the incoming request headers for retrieving the request ID.
+   * @param out reference to a variable where we store the result of the mod operation.
+   * @param mod integer to mod the request ID by.
+   * @return true if request ID is valid and out is populated by the result.
+   */
+  virtual bool modBy(const Http::RequestHeaderMap& request_headers, uint64_t& out,
```
Comment (htuch) [inline include/envoy/http/request_id_extension.h:28]: I don't have the full history here, but it seems that if we're going to make an extension, do we have any value in having every extension do the `%` operation? Should this method really be something like `obtainIntegerRepresentingRequestIdInHeadersToDoSomeModLater`? (Obviously shorter).

File: source/common/stream_info/stream_info_impl.h [hunk 5]
```diff
@@ -253,13 +253,17 @@ struct StreamInfoImpl : public StreamInfo {
 
   const Http::RequestHeaderMap* getRequestHeaders() const override { return request_headers_; }
 
-  void setRequestIDExtension(Http::RequestIDExtensionSharedPtr utils) override {
-    request_id_extension_ = utils;
+  void setRequestIDProvider(const Http::RequestIdStreamInfoProviderSharedPtr& provider) override {
+    ASSERT(provider != nullptr);
```
Comment (htuch) [inline source/common/stream_info/stream_info_impl.h:257]: Is this `ASSERT` or `ENVOY_BUG`?

File: test/extensions/request_id/uuid/config_test.cc [hunk 6]
```diff
@@ -154,29 +161,45 @@ TEST(UUIDRequestIDExtensionTest, DISABLED_benchmark) {
...
+  envoy::extensions::request_id::uuid::v3::UuidRequestIdConfig config;
+  config.mutable_pack_trace_reason()->set_value(false);
+  UUIDRequestIDExtension uuid_utils(config, random);
+
+  Http::TestRequestHeaderMapImpl request_headers;
+  request_headers.setRequestId(random.uuid());
+  EXPECT_EQ(Tracing::Reason::NotTraceable, uuid_utils.getTraceReason(request_headers));
+
+  uuid_utils.setTraceReason(request_headers, Tracing::Reason::Sampling);
+  EXPECT_EQ(Tracing::Reason::NotTraceable, uuid_utils.getTraceReason(request_headers));
```
Comment (htuch) [inline test/extensions/request_id/uuid/config_test.cc:200]: Might have missed it, but is there some test that validates the `modBy` value remains unchanged when packing is disabled?

**Review by htuch — COMMENTED** [review]
Just a few minor comments.

File: include/envoy/http/request_id_extension.h (hunk 4 above)
Comment (euroelessar) [reply include/envoy/http/request_id_extension.h:28]: That would work, though the vast majority of the complexity is still in "obtaining the integer" logic

Comment (mattklein123) [reply include/envoy/http/request_id_extension.h:28]: Agree and will change.

File: test/extensions/request_id/uuid/config_test.cc (hunk 6 above)
Comment (mattklein123) [reply test/extensions/request_id/uuid/config_test.cc:200]: Will update SetTraceStatusPackingDisabled test to explicitly also check that the request ID has not changed.

File: source/common/stream_info/stream_info_impl.h (hunk 5 above)
Comment (mattklein123) [reply source/common/stream_info/stream_info_impl.h:257]: ENVOY_BUG seems unnecessary to me for this one but will switch if you prefer.

**Comment by mattklein123** [comment]
@htuch updated

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**